### PR TITLE
User selection of the workspace namespace/project

### PR DESCRIPTION
**Pull request by metlos**
### What does this PR do?
`$TITLE`

This is a big refactoring that enables us to pass additional information on top of the `RuntimeIdentity` from the workspace API down to the infrastructures.

It then takes advantage of this new class and passes the selection of the namespace to the infrastructure API.

### What issues does this PR fix or reference?
#14908

### CODE REVIEW

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/model/impl/RuntimeTarget.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.workspace.server.model.impl;
+
+import java.util.Objects;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.commons.annotation.Nullable;
+import org.eclipse.che.commons.subject.Subject;
+
+public final class RuntimeTarget {
```
Comment (sleshchenko) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/model/impl/RuntimeTarget.java]: Please add a few lines that would explain what is `runtime target`. For me, it's not obvious...
I understand targetNamespace as namespace where workspace is run...
And I assume that RuntimeTarget is something like who owns runtime, to whom this runtime belongs to...

Comment (metlos) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/model/impl/RuntimeTarget.java]: Hopefully the the javadoc makes it clearer now.

Comment (sleshchenko) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/model/impl/RuntimeTarget.java]: Sorry, for me `RuntimeTarget` is not clear yet... Why `ownerId` is `RuntimeIdentity` and ownerName is part of `RuntimeTarget`?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 2]
```diff
@@ -56,6 +53,7 @@
  */
 @Singleton
 public class KubernetesNamespaceFactory {
```
Comment (mshaposhnik) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java:60]: wow the logic in this class is HUGE!  imho it's time to think how to split it.. maybe using strategy pattern (legacy vs new rules etc)

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 3]
```diff
@@ -323,122 +356,111 @@ private KubernetesNamespaceMeta asNamespaceMeta(Namespace namespace) {
...
-      namespaceName = evalNamespaceName(workspaceId, EnvironmentContext.getCurrent().getSubject());
-    } catch (NotFoundException | ServerException | ConflictException | ValidationException e) {
-      throw new InfrastructureException(
-          format(
-              "Failed to determine the namespace to put the workspace %s to."
-                  + " The error message was: %s",
-              workspaceId, e.getMessage()),
-          e);
-    }
+  public KubernetesNamespace getOrCreate(RuntimeTarget target) throws InfrastructureException {
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java]: the name of this method is confusing. I would expect it to get the namespace if it exists or create a new one and return if it does not. But it looks like it's doing something a bit different. Could you please at least document the method and maybe think of better name?

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/OpenShiftProjectFactory.java [hunk 4]
```diff
@@ -84,67 +76,38 @@ public OpenShiftProjectFactory(
...
-      throw new InfrastructureException(
-          format(
-              "Failed to evaluate the project name to use for workspace %s. The error message"
-                  + " was: %s",
-              workspaceId, e.getMessage()),
-          e);
-    }
-    OpenShiftProject osProject = doCreateProject(workspaceId, projectName);
-    osProject.prepare();
+  public OpenShiftProject getOrCreate(RuntimeTarget target) throws InfrastructureException {
```
Comment (sparkoo) [inline infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/OpenShiftProjectFactory.java]: same as k8s

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java [hunk 5]
```diff
@@ -436,6 +443,21 @@ private void startAsync(
       throw new ConflictException(e.getMessage(), e);
     }
 
+    // handle the situation where a workspace created by a previous Che version doesn't have a
+    // namespace stored for it. In this case, we just store the default namespace for it.
+    if (isNullOrEmpty(
```
Comment (sparkoo) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java:450]: Would it be possible to fix workspace object as soon as possible in the chain, so we would avoid `if`s like this in the code later?

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceService.java [hunk 6]
```diff
@@ -813,7 +813,7 @@ public void deleteProject(
   private static Map<String, String> parseAttrs(List<String> attributes)
       throws BadRequestException, ForbiddenException {
     if (attributes == null) {
-      return emptyMap();
+      return new HashMap<>();
```
Comment (sparkoo) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceService.java]: why this change?

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceValidator.java [hunk 7]
```diff
@@ -51,6 +53,13 @@
       Pattern.compile("[a-zA-Z][a-zA-Z0-9-_.]{0,18}[a-zA-Z0-9]");
   private static final Pattern VOLUME_PATH = Pattern.compile("/.+");
 
+  private Set<WorkspaceAttributeValidator> attributeValidators;
```
Comment (sparkoo) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceValidator.java]: can it be final ?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java (hunk 2 above)
Comment (sparkoo) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java:60]: yes, I lold at this `boolean newLogic =`

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceAttributeValidator.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+public interface WorkspaceAttributeValidator {
+
+  /**
+   * Validates if the specified workspace attributes does not contain invalid attributes according
+   * to implementors rules.
+   *
+   * @param attributes workspace attributes to validate
+   * @throws ValidationException when the specified workspace attributes is not valid
+   */
+  void validate(Map<String, String> attributes) throws ValidationException;
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceAttributeValidator.java:34]: > IMHO exception should handle exceptional state. The failed validation is not exceptional.


Agree and not =)

If Java would be Go we would `isValid, message`
Now we are able to go two ways:
1. Throw exception with a concrete message for a user.
2. Introduce some ValidationResult and two fields `isValid, message`.

I believe from a user perspective of view it's much better to use
`Hey, namespace attribute is longer than 63 characters` or `Hey, you're not permitted to use any namespace, only allowed - `sleshche-che`` instead of `Something is wrong with your workspace/attributes`

Would you go with 2nd approach? 
I'm not sure that with current architecture, it would brings a lot of profit, ValidationResult should be transformed to Exception maybe on another level like WorkspaceManager, WorkspaceService.
I'm not sure that we should invest time to rework like `createWorkspace` to one that returns ValidationResult...
Hope I described my opinion pretty clear - if no - feel free to ask details.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceService.java (hunk 6 above)
Comment (sleshchenko) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceService.java]: I believe it may be reverted after the last refactoring in this area. Will do

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceValidator.java (hunk 7 above)
Comment (sleshchenko) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceValidator.java]: I believe so, will do

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceAttributeValidator.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (sparkoo) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceAttributeValidator.java:34]: yes, I've realized the same after I've submitted the review. 100% for result object. And yes, it might be transformed into the exception where appropriate. Preferably unchecked exception that will be catched where it is appropriate.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java (hunk 5 above)
Comment (metlos) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java:450]: This is the soonest we can do it in a normal flow. We could theoretically do this during workspace recovery, but that would be a much larger change, because workspace recovery currently only reconciles running workspaces.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java (hunk 3 above)
Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java]: The thing is that this will return a namespace and it will also create it if configured to do so. So it will get it or create and get it.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 9]
```diff
@@ -323,122 +356,111 @@ private KubernetesNamespaceMeta asNamespaceMeta(Namespace namespace) {
...
   }
 
-  public void delete(String workspaceId) throws InfrastructureException {
-    final String namespaceName;
-    try {
-      namespaceName = evalNamespaceName(workspaceId, EnvironmentContext.getCurrent().getSubject());
-    } catch (NotFoundException | ServerException | ConflictException | ValidationException e) {
-      throw new InfrastructureException(
-          format("Failed to determine the namespace of the workspace %s", workspaceId), e);
+  public KubernetesNamespace get(RuntimeTarget target) throws InfrastructureException {
```
[the remark this answers is not in the document]
Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java]: why do you think the name is confusing? We are in a namespace factory and we are getting a namespace for given target.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java (hunk 2 above)
Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java:60]: Yes, this whole class is a mess. The naming was much more confusing before I tried to make it at least a little bit more approachable - `create()` would only instantiate an object that is able to access a namespace as just one example. We used the `EnvironmentContext` to get the current user on codepaths that don't originate in the REST API, etc.

I agree we should think about splitting this class. Just not sure we should do it at the same time as changing the logic.

I didn't find `newLogic` too funny when I had to write it to handle back compat ;) Strategy pattern would be good, but that would probably require quite a lot of change. The split between old and new logic is based on existence of a namespace - we cannot detect that in the constructor. Also the split is needed in a couple of methods that are hooked to a much larger workflow way beyond this class.

I guess we have different thresholds for code ugliness, but I still find it better to try and make more sense of this class as it is in the broader picture than try to rearchitect the workspace provisioning.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 10]
```diff
@@ -323,122 +356,111 @@ private KubernetesNamespaceMeta asNamespaceMeta(Namespace namespace) {
...
+  }
+
+  /**
+   * Gets a namespace the workspace is deployed to.
+   *
+   * @param workspace the workspace
+   * @return the namespace
+   */
+  public KubernetesNamespace get(Workspace workspace) throws InfrastructureException {
+    return doCreateNamespaceAccess(workspace.getId(), getNamespaceName(workspace));
```
[the remark this answers is not in the document]
Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java:352]: A `get` method gives you access to a namespace of the workspace. Note that before, this method was called `create` :wink:

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/OpenShiftProjectFactory.java (hunk 4 above)
Comment (metlos) [reply infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/OpenShiftProjectFactory.java]: It isn't. (just trying to win the contest for the shortest PR review comment)

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/model/impl/RuntimeTarget.java (hunk 1 above)
Comment (sleshchenko) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/model/impl/RuntimeTarget.java]: is not actual anymore

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceValidator.java (hunk 7 above)
Comment (sleshchenko) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceValidator.java]: fixed

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceService.java (hunk 6 above)
Comment (sleshchenko) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceService.java]: reverted

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceAttributeValidator.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (skabashnyuk) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceAttributeValidator.java:34]: @sparkoo to make sure I've understand your last comment. Do you want to ValidationResult instead of the exception?

[the remark this answers is not in the document]
Comment (sparkoo) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceAttributeValidator.java:34]: @skabashnyuk yes, I would prefer that.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/PodTerminationGracePeriodProvisioner.java [hunk 11]
```diff
@@ -54,7 +54,7 @@ public void provision(KubernetesEnvironment k8sEnv, RuntimeIdentity identity)
   }
 
   /**
-   * @param pod
+   * @param podSpec
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/PodTerminationGracePeriodProvisioner.java]: if you don't document the param, it doesn't make sense to have it here at all

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceAttributeValidator.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (metlos) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceAttributeValidator.java:34]: -1. Instantiating an object each time even in a hopefully much more common happy case IMHO outweighs the positives.

Java has exceptions for rich handling of exceptional cases. Boolean return type makes this look like C, result object makes this look like Go or Rust. Let's just not.

[the remark this answers is not in the document]
Comment (skabashnyuk) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceAttributeValidator.java:34]: Have the same POV as @sleshchenko and @metlos. I would like to keep ValidationException.

[the remark this answers is not in the document]
Comment (sparkoo) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceAttributeValidator.java:34]: > -1. Instantiating an object each time even in a hopefully much more common happy case IMHO outweighs the positives.

You don't have to instantiate success result. It could be static final instance.

> 
> Java has exceptions for rich handling of exceptional cases. Boolean return type makes this look like C, result object makes this look like Go or Rust. Let's just not.

yes, these newer languages learned from java mistakes.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java [hunk 12]
```diff
@@ -113,14 +113,21 @@ public KubernetesNamespace(
    * <p>Preparing includes creating if needed and waiting for default service account.
    *
    * @param markManaged mark the namespace as managed by Che if it is newly created in this method
+   * @param canCreate true if the namespace can be created during the prepare, false if the
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java]: weird word order in this sentence. Maybe something like "If true, the namespace will be created when does not exist. If false, namespace must already exist otherwise exception will be thrown." ?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java [hunk 13]
```diff
@@ -113,14 +113,21 @@ public KubernetesNamespace(
...
    */
-  boolean prepare(boolean markManaged) throws InfrastructureException {
+  boolean prepare(boolean markManaged, boolean canCreate) throws InfrastructureException {
     KubernetesClient client = clientFactory.create(workspaceId);
     Namespace namespace = get(name, client);
 
     if (namespace == null) {
+      if (!canCreate) {
+        throw new InfrastructureException(
+            format("Creating the namespace '%s' is not allowed, yet" + " it was not found.", name));
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java]: unnecessary concat

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java (hunk 2 above)
Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java:60]: https://github.com/eclipse/che/issues/15250

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesObjectUtil.java [hunk 14]
```diff
@@ -43,6 +43,22 @@
 
   private static final String STORAGE_PARAM = "storage";
 
+  /** Checks if the specified object has the specified label. */
+  public static boolean isLabeled(HasMetadata source, String key, String value) {
```
[the remark this answers is not in the document]
Comment (sparkoo) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesObjectUtil.java:47]: I can't find tests for the method.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/RuntimeInfrastructure.java [hunk 15]
```diff
@@ -90,31 +90,58 @@ public EventService getEventService() {
...
 
+  /**
+   * Returns the namespace a workspace should be deployed into when user do not specify it.
+   *
+   * <p>May be used for evaluating a default namespace or for workspaces that does not have stored
+   * infrastructure namespace info(legacy workspaces).
+   *
+   * @param resolutionCtx the runtime holder specifying which user and workspace runtime targets.
+   * @throws InfrastructureException when there is no configured default namespace
+   * @throws InfrastructureException on any other error
```
Comment (sparkoo) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/RuntimeInfrastructure.java]: same exception thrown twice ?

**Review by sparkoo — APPROVED** [review]
few javadoc nitpicks

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesObjectUtil.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesObjectUtil.java:47]: Throughout the codebase, we tend to mark the parameters that are expected to be null with the `org.eclipse.che.commons.annotation.Nullable` annotation.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java [hunk 16]
```diff
@@ -227,6 +298,19 @@ private void create(String namespaceName, KubernetesClient client)
     }
   }
 
+  private void update(Namespace namespace, KubernetesClient client) throws InfrastructureException {
+    try {
+      client.namespaces().createOrReplace(namespace);
```
Comment (amisevsk) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java:303]: Is `createOrReplace` the correct logic here, or would an exception when namespace doesn't already exist be more appropriate?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java [hunk 17]
```diff
@@ -200,15 +121,37 @@ public boolean isManagingNamespace(String workspaceId) throws InfrastructureExce
...
+    if (allowUserDefinedNamespaces) {
+      // any namespace name is allowed but workspace start may fail
+      return;
+    }
+
+    String defaultNamespace =
+        evalPlaceholders(defaultNamespaceName, EnvironmentContext.getCurrent().getSubject(), null);
+    if (!namespaceName.equals(defaultNamespace)) {
+      throw new ValidationException(
+          "User defined namespaces are not allowed. You're able to specify only admin configured which is '"
```
Comment (amisevsk) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java]: This message could be a little better-worded, something along the lines of 

```java
String.format("User defined namespaces are not allowed. Only the default namespace '%s' is available", defaultNamespaceName) 
```

maybe?

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/NamespaceResolutionContext.java [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.workspace.server.spi;
+
+import java.util.Objects;
+
+/**
+ * Infrastructure may be configured to use namespace value that should be evaluated. This objects
+ * holds everything that might be needed for such evaluation.
```
Comment (amisevsk) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/NamespaceResolutionContext.java]: I would lead with what the class is for ("Holds information needed for namespace name evaluation") and follow up with why in what scenarios it's needed.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java (hunk 16 above)
Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java:303]: This is a private method called from a place where we are sure the namespace already exists and we only update it with new metadata.

There is no "update" on the kubernetes API for namespaces as far as I can tell. Hence the choice of this call.

**Comment by dmytro-ndp** [comment]
@metlos: there are changes in Happy path tests in master branch. Could you, please, take that changes?

**Comment by dmytro-ndp** [comment]
crw-ci-test

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java (hunk 16 above)
Comment (amisevsk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java:303]: Fair enough, I assumed there was a `#replace` or `#update` method available.

**Comment by dmytro-ndp** [comment]
crw-ci-test

**Review by sleshchenko — APPROVED** [review]

**Comment by skabashnyuk** [comment]
ci-test

**Comment by dmytro-ndp** [comment]
crw-ci-test

**Comment by metlos** [comment]
ci-test
